src/adapter/src/catalog/apply.rs

```diff
@@ -1935,6 +1936,72 @@ fn sort_updates_inner(updates: Vec<StateUpdate>) -> Vec<StateUpdate> {
         }
     }
 
+    /// Sort `CONNECTION` items.
+    ///
+    /// `CONNECTION`s can depend on one another, e.g. a `KAFKA CONNECTION` contains
+    /// a list of `BROKERS` and these broker definitions can themselves reference other
+    /// connections. This `BROKERS` list can also be `ALTER`ed and thus it's possible
+    /// for a connection to depend on another with an ID greater than its own.
+    fn sort_connections(connections: &mut Vec<(mz_catalog::durable::Item, Timestamp, StateDiff)>) {
+        let mut topo: BTreeMap<
+            (mz_catalog::durable::Item, Timestamp, StateDiff),
+            BTreeSet<CatalogItemId>,
+        > = BTreeMap::default();
+        let existing_connections: BTreeSet<_> = connections.iter().map(|item| item.0.id).collect();
+
+        // Initialize our set of topological sort.
+        tracing::info!(?connections, "sorting connections");
+        for (connection, ts, diff) in connections.drain(..) {
+            let statement = mz_sql::parse::parse(&connection.create_sql)
+                .expect("valid CONNECTION create_sql")
+                .into_element()
+                .ast;
+            let mut dependencies = mz_sql::names::dependencies(&statement)
+                .expect("failed to find dependencies of CONNECTION");
+            // Be defensive and remove any possible self references.
+            dependencies.remove(&connection.id);
+            // It's possible we're applying updates to a connection where the
+            // dependency already exists and thus it's not in `connections`.
+            dependencies.retain(|dep| existing_connections.contains(dep));
+
+            // Be defensive and ensure we're not clobbering any items.
+            assert_none!(topo.insert((connection, ts, diff), dependencies));
+        }
+        tracing::info!(?topo, ?existing_connections, "built topological sort");
+
+        // Do a topological sort, pushing back into the provided Vec.
+        while !topo.is_empty() {
```

Comment:
This looks like a quadratic loop if you have a chain of connections that depend on each other when this can be linear. 

You can get a better algorithm by constructing a `todo` vector that initially contains only items with no dependencies, which you find from above. Then in this loop you repeatedly pop an item from the todo, push any dependencies it might have back in todo with a single lookup, and then push the item back in `connections`.

Reply:
do we have any data on how large the set of connections gets in practice? can we do the algorithm improvements in a follow up? no totally sure how broken flare is in prod right now either.

Reply:
It's a simple change to make it linear so I don't think it makes sense to do a follow up with another round of review and a chance of never happening. It's just BFS :)

Reply:
fair point

Reply:
@ptravers is interested in improving this algo, to get folks unblocked I'm going to merge this so I can cut a patch release :)